### ISSUE TITLE
masked convolution for convolutional networks that consider HOP

### ISSUE THREAD

**Issue by navid-mahmoudian** [rusty1s/pytorch_geometric#1205]
## 🚀 Feature
Hello, 
Thank you for providing this great library. Is it possible to provide a masked version of the convolutional networks that consider HOP, i.e. ChebConv, TAGConv, SGConv?

Let's say that the module accepts parameters of each of these convolutional networks plus a mask that indicates the center nodes that we want to compute convolution around them. Note that the neighboring nodes **can be outside this mask**. That's why simple cutting the graph is not working because we want to use also neighboring nodes that are not in the mask in the message passing algorithm.

I see that you have already implemented  DenseGCNConv,  DenseSAGEConv, and  DenseGraphConv, but :
1) first these convolutional networks only consider 1 hop neighborhood.
2) I don't know if even in this one-hop neighborhood they consider neighbor nodes that are not in the mask.

**Comment by rusty1s** [comment]
If I understand you correctly, you are looking for a way to get the L-Hop subgraph around masked nodes? You can do this with the [`k_hop_subgraph](https://pytorch-geometric.readthedocs.io/en/latest/modules/utils.html#torch_geometric.utils.k_hop_subgraph) utility function.

**Comment by navid-mahmoudian** [comment]
And then apply convolution around each node in a for loop? Is it efficient in terms of performance?

**Comment by rusty1s** [comment]
I see, `k_hop_subgraph` currently only works for a single node. It should be no problem to extend this to multiple nodes.

**Comment by navid-mahmoudian** [comment]
I don't think that it is efficient, because finding subgraph is already I have to do it per node, and then applying convolution on each node would cause the running time to increase drastically.
I just want in the convolution to choose the central nodes to apply the convolution around them.

**Comment by navid-mahmoudian** [comment]
> I see, `k_hop_subgraph` currently only works for a single node. It should be no problem to extend this to multiple nodes.

Yes, exactly, you answered faster. But also I want in the convolution to choose the central nodes to apply the convolution around them, but as I said, in the message passing of convolution it is possible to pass messages from the nodes which are not in the mask.

**Comment by rusty1s** [comment]
Well, i see no reason to put that behaviour inside each convolution when it is basically just a modified input. Note that `k_hop_subgraph` now supports passing in multiple nodes in master (either via list, tuple, or tensor).

**Comment by navid-mahmoudian** [comment]
Ok. I see. 
Thank you very much

**Comment by navid-mahmoudian** [comment]
Just a question:
Let's say I have n nodes in the mask (out of m nodes of the whole graph with n < m). I want to apply TAGConv with K=4 on these n nodes while the whole graph has m nodes. I don't understand how we can do this even now that k_hop_subgraph supports multiple nodes

**Comment by rusty1s** [comment]
```python
subset, sub_edge_index, mapping, _ = k_hop_subgraph(node_idx, edge_index,
                                                    k=4, relabel_nodes=True)
h= F.relu(conv1(x[subset], sub_edge_index))
...
h= F.relu(convx(x[subset], sub_edge_index))
out = h[mapping]
```

**Comment by navid-mahmoudian** [comment]
Maybe, I have not explained clearly what I want. Imagine this simple program:
```

import torch
import torch_geometric

B = 1  # batch size
in_channels = 1
out_channels = 8
K = 2   # Number of hops
node_dim = 1    # data is in form [batch_size, num_nodes, num_features]
conv = torch_geometric.nn.TAGConv(in_channels=in_channels, out_channels=out_channels, K=K, node_dim=node_dim) # Graph convolution: 'ChebConv', 'TAGConv', 'SGConv',

edge_index = torch.tensor([[0, 1, 3, 2],
                           [1, 3, 2, 0]])
N= edge_index.max().item() + 1

interested_nodes = torch.tensor([1,2])
subset, sub_edge_index, mapping, _ = torch_geometric.utils.k_hop_subgraph(interested_nodes, edge_index=edge_index, num_hops=K, relabel_nodes=True)

x= torch.randn(B, N, in_channels)

# print("x=", x)

out1 = conv(x, edge_index=edge_index)
print("conv at interested nodes=", out1.index_select(node_dim, interested_nodes))
out2 = conv(x.index_select(node_dim, subset), sub_edge_index)
print("out2=", out2)
```

Basically what I want is that I want to have **out1.index_select(node_dim, interested_nodes)**, but without doing the convolution on the whole x because x is very big. However, the solution that you gave not only doesn't give the same answer but also the size doesn't match.

**Comment by rusty1s** [comment]
For me, the output is the same. Note that you again need to filter `out2` to only contain the important nodes (not the ones that were used directly in message passing):
```python
out2 = out2.index_select(node_dim, mapping)
```

**Comment by navid-mahmoudian** [comment]
Thank you for your answer. Maybe my example was very easy because then hop 2 contains the whole graph. In the following code which I used the same code but for a more complex graph as below
![graph_two_hexagon](https://user-images.githubusercontent.com/10260746/81913223-f33f4400-95cf-11ea-9e37-6530cf5da2ee.jpg)

the answer is not the same. Let's assume the masked nodes are nodes 1,2. Note that since I used random generation for input x each time the difference is different, but sometimes the difference is as big as 0.03 between each element:

```
import torch
import torch_geometric
import itertools
import numpy as np


B = 1  # batch size
in_channels = 1
out_channels = 8
K = 2   # Number of hops
node_dim = 1    # data is in form [batch_size, num_nodes, num_features]
conv = torch_geometric.nn.TAGConv(in_channels=in_channels, out_channels=out_channels, K=K, node_dim=node_dim) # Graph convolution: 'ChebConv', 'TAGConv', 'SGConv',

# create a graph which consists of 2 hexagone which are linked with one edge between not 0 and 9
edge_index = list(itertools.permutations(range(0,6), 2)) + list(itertools.permutations(range(6,12), 2)) + [(0,9), (9,0)]
edge_index = np.transpose(edge_index)
edge_index = torch.from_numpy(edge_index)
# print("edge_index=", edge_index)
N= edge_index.max().item() + 1

interested_nodes = torch.tensor([1,2])
subset, sub_edge_index, mapping, _ = torch_geometric.utils.k_hop_subgraph(interested_nodes, edge_index=edge_index, num_hops=K, relabel_nodes=True)

x= torch.randn(B, N, in_channels)

# print("x=", x)
# print("subset=", subset)
# print("sub_edge_index=", sub_edge_index)

out1 = conv(x, edge_index=edge_index)
# print("out1 (before)=", out1)
out1 = out1.index_select(node_dim, interested_nodes)

out2 = conv(x.index_select(node_dim, subset), sub_edge_index)
# print("out2 (before)=", out2)
out2 = out2.index_select(node_dim, mapping)
print("diff (out2-out1)=", out2-out1)
# print("out1=", out1)
# print("out2=", out2)
```

**Comment by rusty1s** [comment]
I found the issue. Not sure if it is a bug though. `TAGConv`/`GCNConv` normalization depends on the node degree and the degrees of its neighbors. In the subgraph, the degree of neighbors is a different one. I therefore added the `normalize` argument to `TAGConv` (which was already present in `GCNConv`) so that you can now precompute normalization coefficients and pass them to the operator:
```python
out1 = conv(x, edge_index=edge_index)

conv.normalize = False
edge_index, edge_weight = torch_geometric.nn.TAGConv.norm(edge_index, num_nodes=N)

interested_nodes = torch.tensor([1, 2])
subset, sub_edge_index, mapping, edge_mask = torch_geometric.utils.k_hop_subgraph(interested_nodes, edge_index, num_hops=K, relabel_nodes=True)

out2 = conv(x.index_select(node_dim, subset), sub_edge_index, edge_weight[edge_mask])

out1 = out1.index_select(node_dim, interested_nodes)
out2 = out2.index_select(node_dim, mapping)
print("diff (out2-out1)=", out2 - out1)
```
